source/vistas/core/graphics/factory.py

```diff
@@ -0,0 +1,123 @@
+import mercantile
+
+from vistas.core.bounds import union_bboxs
+from vistas.core.graphics.bounding_box import BoundingBoxHelper
+from vistas.core.graphics.objects import Object3D
+from vistas.core.task import Task
+from vistas.core.threading import Thread
+from vistas.ui.utils import post_redisplay
+
+
+class MeshFactoryWorker(Thread):
+    """ Interface for executing work from a MeshFactory in a separate thread. """
+
+    task_name = "Creating Meshes"
+    task_description = task_name
+
+    def __init__(self, factory, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.factory = factory
+        self.task = Task(self.task_name, self.task_description)
+
+    def work(self):
+        """ The work that should be done by this worker thread. """
+
+        raise NotImplementedError
+
+    def run(self):
+        self.task.status = Task.RUNNING
+        self.init_event_loop()
+        self.work()
+        self.sync_with_main(post_redisplay)
+        self.task.status = Task.COMPLETE
+
+
+class MeshFactory(Object3D):
```

Comment:
That is true, perhaps what gets emitted from a `MeshFactory` is a `MeshGroup` which implements the methods from `Object3D` and the `render` method necessary to be rendered. The `MeshFactory` would need to know about the `Scene` that the `MeshGroup` belongs to though, which complicates the interaction with the plugins a bit. Worth testing out.

Reply:
Is a `MeshFactory` even necessary? This class already seems to be a `MeshGroup`, so it's probably fine to rename it and otherwise leave as-is.

Reply:
I think that it may be good to have them separate. A `MeshGroup` could be a group of simple meshes, while this `MeshFactory` is what is actually creating the group. A `MeshGroup` could be grass or trees or other 3D features, something that our collaborators have talked about.

That said, I'm fine with calling it a `MeshGroup` for now, and if we need to break it down more granular, then we can do so.

Reply:
I guess I'm seeing how it's a collection of meshes and handles rendering... I don't see where it _creates_ them.

Reply:
Mesh geometry (and potentially other long-running pieces) are created in the factory's `MeshWorkerThread`, using the `build` method. The `MeshFactory` starts that factory's `MeshFactoryWorker` thread, which is where the long running geometry creation process happens, which happens in the `work` method. Then the worker communicates back to the factory what should happen with that geometry.

For the `TerrainTileFactory`, it's worker just tosses the geometry back to the factory and the factory adds a `Mesh` to it's `items` attribute, which signals that it can now be rendered into the scene. When the worker is done with all it's assigned work, the `run` method does a refresh on the screen. Having the `run` method use the internal `work` method is a consistent way to make sure that the `event_loop` is started, which prevents problems with the `ElevationSource` throwing errors because no event loop was started in the worker thread.